<a href="https://colab.research.google.com/github/Wipada6430205202/BSC_DPDM23/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import os

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย Groupby.apply(list)

In [ ]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
data.shape

(541909, 8)

###Quiz หาว่ามีกี่ประเทศ

In [ ]:
data['Country'].value_counts().shape

(38,)

## Association Rules

In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=b4a296898e6c79b879c6a0c8c0f5540895013e439f5d2604a58e136657c54c4e
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
from apyori import apriori

In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [ ]:
len(frequentItemsets)

4

In [ ]:
frequentItemsets[0][0]

frozenset({'22423'})

In [ ]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [ ]:
frequentItemsets[0][0].__len__()

1

In [ ]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [ ]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'562717',
 '543045',
 '564817',
 '561637',
 '566193',
 '568060',
 '564092',
 '551997',
 '558862',
 '571713',
 '537240',
 '542619',
 '567293',
 '537396',
 '579187',
 '549222',
 '564835',
 '555722',
 '546136',
 '558120',
 'C539071',
 '578817',
 '575703',
 '578067',
 '576202',
 '558116',
 '546660',
 '571034',
 '539594',
 '539609',
 '548987',
 '560246',
 'C545033',
 '545428',
 '549951',
 '560473',
 'C545438',
 '546896',
 '556206',
 '562711',
 '553891',
 '540499',
 '543107',
 '573585',
 '553035',
 '555851',
 '562282',
 '565150',
 '541960',
 '545411',
 '564191',
 '554013',
 '541584',
 '545216',
 '553542',
 'C579945',
 '563166',
 '545895',
 '552956',
 '559055',
 '538566',
 '550986',
 '562607',
 '551387',
 '548260',
 '542349',
 '569834',
 '567671',
 '552867',
 '571061',
 '572854',
 '538370',
 '559817',
 '552563',
 '545214',
 '552694',
 '561891',
 '539735',
 '544660',
 '543989',
 '573020',
 '576246',
 '537642',
 '555106',
 '573388',
 'C555409',
 '560826',
 '564057',
 '537179',
 '565396',
 '564

In [ ]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

### Quiz 22386 และ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [ ]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Homework1 หา Frequent Item Set และ Assosiation Rules
-  ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec) [หนามเตย]
-  ของประเทศในทวีป Europe [จจ]
-  ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb) [ใบเตย]
-  ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec) [พิม]
-  ของวันวาเลนไทน์ [ตุนนี่]
-  โดยจัดกลุ่มตามประเทศ+เดือน [พิมจ๊อก]

In [ ]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##จัดการข้อมูล

แยกวัน เดือน ปี เวลา

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Day  Month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    1     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    1     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   

   Year      Time  
0  2010  08:26:00  
1  2010  08:26:00  
2  2010  08:26:00  
3  2010  08:26:00  
4  2010  0

##1.ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec)

In [ ]:
import pandas as pd
import os
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/Mining'

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/Mining/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# แปลงคอลัมน์ timestamp เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# แยกวันที่เป็นเดือน, วัน, และปีแยกกัน
data['month'] = data['InvoiceDate'].dt.month
data['day'] = data['InvoiceDate'].dt.day
data['year'] = data['InvoiceDate'].dt.year


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
data

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  month  day  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom     12    1   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     12    1   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom     12    1   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     12    1   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom     12    1   
...                    ...        ...         ...             ...    ...  ...   
541904 2011-12-09 12:50:00       0.85     12680.0          France     12    9   
541905 2011-12-09 12:50:00       2.10     12680.0          France     12    9   
541906 2011-12-09 12:50:00       4.15     12680.0          France     12    9   
541907 2011-12-09 12:50:00       4.15     12680.0          France     12    9   
541908 2011-12-09 12:50:00       4.95     12680.0          France     12    9   

        year  
0       2010  
1       2010  
2       2010  
3       2010  
4       2010  
...      ...  
541904  2011  
541905  2011  
541906  2011  
541907  2011  
541908  2011  

[541909 rows x 11 columns]

In [ ]:
# แปลงข้อมูลเป็น Transaction List
chrismas_and_newyear_transactions= data.groupby('InvoiceNo')['StockCode'].apply(list)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
!pip install apyori
from apyori import apriori

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
chrismas_and_newyear_transactions

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [ ]:
min_support = 0.03
christmas_and_newyear_frequentItemsets = list(apriori(chrismas_and_newyear_transactions, min_support=min_support))
for i in christmas_and_newyear_frequentItemsets :
   if i[0].__len__()>1:
    print('Assosiation Rules')
    print(f'frequentItemsets is{i[0]}')
    print(f'have support{i[1]}')
    print(f'othar {i[2:]}')


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Assosiation Rules
frequentItemsets isfrozenset({'85099B', '22386'})
have support0.03216216216216216
othar ([OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)],)
Assosiation Rules
frequentItemsets isfrozenset({'22699', '22697'})
have support0.03027027027027027
othar ([OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.152317880794

##2.ของประเทศในทวีป Europe

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

In [ ]:
set(data['Country'])

In [ ]:
EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [ ]:
EU_data_country = data[data['Country']==""]
for i in range(len(EU_country)):
  x = data[data['Country']==EU_country[i]]
  EU_data_country=pd.concat([x,EU_data_country])

In [ ]:
EU_data_country

In [ ]:
EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
transactions = EU_data_country.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori


In [ ]:
apriori(transactions,min_support = 0.053)


In [ ]:
list(apriori(transactions,min_support = 0.053))


In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.053))
frequentItemsets

In [ ]:
len(frequentItemsets)


In [ ]:
frequentItemsets[0]


In [ ]:
frequentItemsets[0][0]

In [ ]:
dir(frequentItemsets[0][0])

In [ ]:
frequentItemsets[0][0].__len__()


In [ ]:
min_sup = 0.032

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='85099B')]['Description'])


In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='22807')]['Description'])


In [ ]:
transactions =EU_data_country.groupby(['Country', 'Year'])['StockCode'].apply(list)
min_sup = 0.4
frequentItemsets= list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(EU_data_country[(EU_data_country['StockCode']=='22423')]['Year'])


In [ ]:
EU_data_country[EU_data_country['StockCode']=='22252']

In [ ]:
EU_data_country[EU_data_country['InvoiceNo']=='543235']

##3.ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb)

In [ ]:
import pandas as pd
import os
!pip install apyori
from apyori import apriori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=e73f9e3f97eee7a7aecef33a20682ac4df2ed2d1260fe80a5af59dfda7276f37
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

Mounted at /content/drive


In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time


In [ ]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') |
            (data['Country']=='Lebanon') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]
Asia #แบ่งทวีปเอเชีย


InvoiceNo StockCode                     Description  Quantity  \
9783      537218     85016   SET OF 6 VINTAGE NOTELETS KIT         6   
9784      537218     21506      FANCY FONT BIRTHDAY CARD,         12   
9785      537218     22716              CARD CIRCUS PARADE        12   
9786      537218     22983             CARD BILLBOARD FONT        12   
9787      537218     22030          SWALLOWS GREETING CARD        12   
...          ...       ...                             ...       ...   
530276   C580832     21731   RED TOADSTOOL LED NIGHT LIGHT        -2   
530277   C580832     23480  MINI LIGHTS WOODLAND MUSHROOMS       -15   
530278   C580832     22466  FAIRY TALE COTTAGE NIGHT LIGHT       -17   
530279   C580832     23084              RABBIT NIGHT LIGHT        -7   
530280   C580832     22621      TRADITIONAL KNITTING NANCY        -1   

               InvoiceDate  UnitPrice  CustomerID Country  Day  Month  Year  \
9783   2010-12-05 15:46:00       2.55     12763.0   Japan    5     12  2010   
9784   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9785   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9786   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
9787   2010-12-05 15:46:00       0.42     12763.0   Japan    5     12  2010   
...                    ...        ...         ...     ...  ...    ...   ...   
530276 2011-12-06 11:40:00       1.45     12753.0   Japan    6     12  2011   
530277 2011-12-06 11:40:00       3.29     12753.0   Japan    6     12  2011   
530278 2011-12-06 11:40:00       1.65     12753.0   Japan    6     12  2011   
530279 2011-12-06 11:40:00       1.79     12753.0   Japan    6     12  2011   
530280 2011-12-06 11:40:00       1.45     12753.0   Japan    6     12  2011   

            Time  
9783    15:46:00  
9784    15:46:00  
9785    15:46:00  
9786    15:46:00  
9787    15:46:00  
...          ...  
530276  11:40:00  
530277  11:40:00  
530278  11:40:00  
530279  11:40:00  
530280  11:40:00  

[1314 rows x 12 columns]

In [ ]:
# สร้าง DataFrame สำหรับเหตุการณ์ Chinese New Year
China = Asia[((Asia['Month'] == 1) & (Asia['Day'] >= 15) & (Asia['Day'] <= 31)) |
            ((Asia['Month'] == 2) & (Asia['Day'] <= 15))]

print(China)

      InvoiceNo StockCode                          Description  Quantity  \
69623    541975     22620          4 TRADITIONAL SPINNING TOPS        60   
69624    541975    85183B  CHARLIE & LOLA WASTEPAPER BIN FLORA         6   
69625    541975    85183A   CHARLIE & LOLA WASTEPAPER BIN BLUE         6   
69626    541975     16236                 KITTY PENCIL ERASERS        28   
69627    541975     21058               PARTY INVITES WOODLAND        12   
...         ...       ...                                  ...       ...   
89595    543911    47590B          PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117     METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567            AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910                  WAY OUT METAL SIGN         12   
89599    543911     21911                   GARDEN METAL SIGN         12   

              InvoiceDate  UnitPrice  CustomerID               Country  Day  \
69623 20

In [ ]:
trans = China.groupby('InvoiceNo')['Description'].apply(list).reset_index()
print(trans)

  InvoiceNo                                        Description
0    541975  [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    542102  [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    542276  [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    543179  [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    543518  [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    543911  [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...


In [ ]:
trans_apply = transactions['Description'].apply(lambda x: [str(item) for item in x])
trans_apply

0    [4 TRADITIONAL SPINNING TOPS, CHARLIE & LOLA W...
1    [GIN & TONIC DIET GREETING CARD , GREEN REGENC...
2    [LAUNDRY 15C METAL SIGN, BEWARE OF THE CAT MET...
3    [MONEY BOX FIRST ADE DESIGN, MONEY BOX KINGS C...
4    [SET OF 3 NOTEBOOKS IN PARCEL, LUNCH BAG DOLLY...
5    [RETROSPOT HEART HOT WATER BOTTLE, DOORMAT UNI...
Name: Description, dtype: object

In [ ]:
freqItem = list(apriori(trans_apply, min_support=0.33))

for record in freqItem:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

yes
Items: BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> LUNCH BAG RED RETROSPOT : Confidence=1.0000, Lift=3.0000)
Rule 3: (LUNCH BAG RED RETROSPOT -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=1.0000, Lift=3.0000)
Rule 3: (SET 3 RETROSPOT TEA,COFFEE,SUGAR -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: BATH BUILDING BLOCK WORD, JAM MAKING SET WITH JARS
Support: 0.3333 (33.33%)
Rule 1: (None -> BATH BUILDING BLOCK WORD, JAM MAKING SET WITH JARS : Confidence=0.3333, Lift=1.0000)
Rule 2: (BATH BUILDING BLOCK WORD -> JAM MAKING SET

##4.ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec)

In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Filter data for the Christmas & New Year period and European countries
european_countries = ['Austria', 'Belgium', 'Cyprus', 'Channel Islands', 'Czech Republic', 'Denmark', 'EIRE', 'European Community',
                      'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy', 'Lithuania', 'Malta', 'Netherlands',
                      'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden', 'Switzerland', 'United Kingdom']

filtered_XmasAndNY = data[(data['Month'] == 12) & (data['Country'].isin(european_countries))]

# Create a binary DataFrame
basket = (filtered_XmasAndNY.groupby(['InvoiceNo', 'StockCode'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

basket_sets = basket.applymap(lambda x: 1 if x > 0 else 0)

# Run Apriori algorithm to find frequent item sets
frequent_itemsets = apriori(basket_sets, min_support=0.03, use_colnames=True)

# Display frequent item sets with more than one item
for itemset in frequent_itemsets[frequent_itemsets['itemsets'].apply(lambda x: len(x) > 1)].itertuples(index=False):
    print('yes')
    print(itemset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


yes
Pandas(support=0.04790221341261976, itemsets=frozenset({'22086', '22910'}))
yes
Pandas(support=0.03303600925008259, itemsets=frozenset({'22112', '22111'}))
yes
Pandas(support=0.036669970267591674, itemsets=frozenset({'22112', '22835'}))
yes
Pandas(support=0.036339610175090846, itemsets=frozenset({'22726', '22727'}))
yes
Pandas(support=0.030393128510075983, itemsets=frozenset({'22865', '22866'}))
yes
Pandas(support=0.030393128510075983, itemsets=frozenset({'22865', '22867'}))


In [ ]:
# Coding find Product(Descripyion) with StockCode
def Product(x,y) :
  item =set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return  item

# Create a summary table for comparing two products
product_pairs = [('22086', '22910'), ('22112', '22111'), ('22835', '22112'),
                 ('22726', '22727'), ('22865', '22866'), ('22865', '22867')]

# Create a DataFrame from the summary data
summary_df = pd.DataFrame(columns=['Product Pair', 'Support', 'Product 1 Description', 'Product 2 Description'])

# Populate the DataFrame
for product_pair in product_pairs:
    support = frequent_itemsets[
        frequent_itemsets['itemsets'] == frozenset(product_pair)]['support'].values[0]
    descriptions = Product(*product_pair)

    summary_df = summary_df.append({
        'Product Pair': product_pair,
        'Support': support,
        'Product 1 Description': descriptions.pop(),
        'Product 2 Description': descriptions.pop()
    }, ignore_index=True)

# Display the summary DataFrame
summary_df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:1

Product Pair   Support            Product 1 Description  \
0  (22086, 22910)  0.047902  PAPER CHAIN KIT 50'S CHRISTMAS    
1  (22112, 22111)  0.033036       CHOCOLATE HOT WATER BOTTLE   
2  (22835, 22112)  0.036670       CHOCOLATE HOT WATER BOTTLE   
3  (22726, 22727)  0.036340        ALARM CLOCK BAKELIKE RED    
4  (22865, 22866)  0.030393                              NaN   
5  (22865, 22867)  0.030393          HAND WARMER BIRD DESIGN   

               Product 2 Description  
0  PAPER CHAIN KIT VINTAGE CHRISTMAS  
1       SCOTTIE DOG HOT WATER BOTTLE  
2    HOT WATER BOTTLE I AM SO POORLY  
3         ALARM CLOCK BAKELIKE GREEN  
4             HAND WARMER OWL DESIGN  
5                                NaN

##5.ของวันวาเลนไทน

In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/new/reduced_marketbasket.csv.crdownload'),encoding='latin-1')

In [ ]:
data

แยกวัน

In [ ]:
data['InvoiceDate'][0].split()[0]

In [ ]:
date = []
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)

In [ ]:
date[0].split('/')

['12', '01', '2010']

In [ ]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

นำ data ใหม่ไปใช้

In [ ]:
data_test_nonclean = data

In [ ]:
data_test_nonclean['date'] = date
data_test_nonclean['day'] = day
data_test_nonclean['month'] = month
data_test_nonclean['year'] = year

In [ ]:
set(data_test_nonclean['date'])

{'01/04/2011',
 '01/05/2011',
 '01/06/2011',
 '01/07/2011',
 '01/09/2011',
 '01/10/2011',
 '01/11/2011',
 '01/12/2011',
 '02/01/2011',
 '02/02/2011',
 '02/03/2011',
 '02/04/2011',
 '02/06/2011',
 '02/07/2011',
 '02/08/2011',
 '02/09/2011',
 '02/10/2011',
 '02/11/2011',
 '03/01/2011',
 '03/02/2011',
 '03/03/2011',
 '03/04/2011',
 '03/06/2011',
 '03/07/2011',
 '03/08/2011',
 '03/09/2011',
 '03/10/2011',
 '03/11/2011',
 '04/01/2011',
 '04/03/2011',
 '04/04/2011',
 '04/05/2011',
 '04/06/2011',
 '04/07/2011',
 '04/08/2011',
 '04/10/2011',
 '04/11/2011',
 '04/12/2011',
 '05/01/2011',
 '05/03/2011',
 '05/04/2011',
 '05/05/2011',
 '05/06/2011',
 '05/08/2011',
 '05/09/2011',
 '05/10/2011',
 '05/11/2011',
 '05/12/2011',
 '06/01/2011',
 '06/02/2011',
 '06/03/2011',
 '06/05/2011',
 '06/06/2011',
 '06/07/2011',
 '06/08/2011',
 '06/09/2011',
 '06/10/2011',
 '06/12/2011',
 '07/01/2011',
 '07/03/2011',
 '07/04/2011',
 '07/05/2011',
 '07/06/2011',
 '07/07/2011',
 '07/08/2011',
 '07/10/2011',
 '07/11/20

In [ ]:
set(data_test_nonclean['day'])

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

แปลงวัน

In [ ]:
x=[]
for i in range(1,32):
  x.append(i)
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [ ]:
data_test0 = data_test_nonclean.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                         '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22',
                                         '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)

In [ ]:
data_test_nonclean.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010  01   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010  01   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010  01   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

In [ ]:
data_test0.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country        date  day  \
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  12/01/2010    1   
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  12/01/2010    1   
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   
4  12/01/2010 08:26       3.39     17850.0  United Kingdom  12/01/2010    1   

  month  year  
0    12  2010  
1    12  2010  
2    12  2010  
3    12  2010  
4    12  2010

แปลงเดือน

In [ ]:
set(data_test0['month'])

{1,
 '1',
 10,
 11,
 12,
 2,
 '2',
 3,
 '3',
 4,
 '4',
 5,
 '5',
 6,
 '6',
 7,
 '7',
 8,
 '8',
 9,
 '9'}

In [ ]:
data_test = data_test0.replace(to_replace=["1", "2",'3','4','5','6','7','8','9'],value=[1,2,3,4,5,6,7,8,9])

In [ ]:
data_test.isnull().any()

InvoiceNo      False
StockCode      False
Description     True
Quantity       False
InvoiceDate    False
UnitPrice      False
CustomerID      True
Country        False
date           False
day            False
month          False
year           False
dtype: bool

In [ ]:
data_test_nonclean[data_test_nonclean['month']=='01']

InvoiceNo StockCode                Description  Quantity  \
42481    539993     22386    JUMBO BAG PINK POLKADOT        10   
42482    539993     21499         BLUE POLKADOT WRAP        25   
42483    539993     21498        RED RETROSPOT WRAP         25   
42484    539993     22379   RECYCLING BAG RETROSPOT          5   
42485    539993     20718  RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                        ...       ...   
55385    540992     21912   VINTAGE SNAKES & LADDERS         2   
55386    540992     51008   AFGHAN SLIPPER SOCK PAIR        10   
55387    540992     21508    VINTAGE KID DOLLY CARD         12   
55388    540992     22716         CARD CIRCUS PARADE        12   
55389    540992     22983        CARD BILLBOARD FONT        12   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
55385  01/12/2011 16:57       3.75     17364.0  United Kingdom  01/12/2011   
55386  01/12/2011 16:57       3.45     17364.0  United Kingdom  01/12/2011   
55387  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55388  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   
55389  01/12/2011 16:57       0.42     17364.0  United Kingdom  01/12/2011   

      day month  year  
42481  04    01  2011  
42482  04    01  2011  
42483  04    01  2011  
42484  04    01  2011  
42485  04    01  2011  
...    ..   ...   ...  
55385  12    01  2011  
55386  12    01  2011  
55387  12    01  2011  
55388  12    01  2011  
55389  12    01  2011  

[12909 rows x 12 columns]

In [ ]:
data_test[data_test['month']==1]

InvoiceNo StockCode                    Description  Quantity  \
42481    539993     22386        JUMBO BAG PINK POLKADOT        10   
42482    539993     21499             BLUE POLKADOT WRAP        25   
42483    539993     21498            RED RETROSPOT WRAP         25   
42484    539993     22379       RECYCLING BAG RETROSPOT          5   
42485    539993     20718      RED RETROSPOT SHOPPER BAG        10   
...         ...       ...                            ...       ...   
77623    542745     22138  BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034   REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969   HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649   STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649   STRAWBERRY FAIRY CAKE TEAPOT         1   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
42481  01/04/2011 10:00       1.95     13313.0  United Kingdom  01/04/2011   
42482  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42483  01/04/2011 10:00       0.42     13313.0  United Kingdom  01/04/2011   
42484  01/04/2011 10:00       2.10     13313.0  United Kingdom  01/04/2011   
42485  01/04/2011 10:00       1.25     13313.0  United Kingdom  01/04/2011   
...                 ...        ...         ...             ...         ...   
77623   1/31/2011 16:47       4.95     15881.0  United Kingdom   1/31/2011   
77624   1/31/2011 16:48       0.95     15881.0  United Kingdom   1/31/2011   
77625   1/31/2011 16:53       1.45     15881.0  United Kingdom   1/31/2011   
77626   1/31/2011 16:57       4.95     15881.0  United Kingdom   1/31/2011   
77627   1/31/2011 16:59       4.95     15881.0  United Kingdom   1/31/2011   

       day  month  year  
42481    4      1  2011  
42482    4      1  2011  
42483    4      1  2011  
42484    4      1  2011  
42485    4      1  2011  
...    ...    ...   ...  
77623   31      1  2011  
77624   31      1  2011  
77625   31      1  2011  
77626   31      1  2011  
77627   31      1  2011  

[35147 rows x 12 columns]

In [ ]:
M14_data = data_test[(data_test['month']==2)]

In [ ]:
va_data=M14_data[M14_data['day']==14]

In [ ]:
va_data

InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

           InvoiceDate  UnitPrice  CustomerID         Country       date  day  \
88847   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011   14   
88848   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011   14   
88849   2/14/2011 9:24       5.95     17667.0  United Kingdom  2/14/2011   14   
88850   2/14/2011 9:24       0.42     17667.0  United Kingdom  2/14/2011   14   
88851   2/14/2011 9:24       4.95     17667.0  United Kingdom  2/14/2011   14   
...                ...        ...         ...             ...        ...  ...   
89950  2/14/2011 15:43       1.25     14606.0  United Kingdom  2/14/2011   14   
89951  2/14/2011 15:43       0.85     14606.0  United Kingdom  2/14/2011   14   
89952  2/14/2011 15:43       1.45     14606.0  United Kingdom  2/14/2011   14   
89953  2/14/2011 15:43       6.95     14606.0  United Kingdom  2/14/2011   14   
89954  2/14/2011 16:39       3.75     14189.0  United Kingdom  2/14/2011   14   

       month  year  
88847      2  2011  
88848      2  2011  
88849      2  2011  
88850      2  2011  
88851      2  2011  
...      ...   ...  
89950      2  2011  
89951      2  2011  
89952      2  2011  
89953      2  2011  
89954      2  2011  

[1108 rows x 12 columns]

จัดกลุ่มตาม Transaction

In [ ]:
transactions_va = va_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
len(transactions_va)

55

In [ ]:
min_sup = 0.45
frequentItemsets_va= list(apriori(transactions_va,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

จัดกลุ่มตาม Transaction (ประเทศ + เดือน)

In [ ]:
transactions_va

InvoiceNo
543821     [22926, 22928, 22925, 85175, 22456, 22854, 227...
543822     [22139, 20973, 22585, 22960, 84380, 84378, 229...
543823                          [22138, 21658, 22567, 20914]
543824     [22745, 22746, 20971, 20972, 16225, 22748, 227...
543825     [22926, 22782, 22411, 20914, 22381, 20681, 22690]
543826     [90104, 90146, 90175A, 90181B, 90214J, 90214S,...
543827                                              [84406B]
543828     [22723, 22722, 22720, 22899, 22567, 84987, 228...
543829     [21238, 21239, 21240, 84997B, 84997D, 84997C, ...
543831     [84519A, 20711, 20713, 85099F, 22385, 21929, 2...
543832     [84879, 22077, 22855, 21899, 21900, 21901, 219...
543833     [22788, 22957, 21621, 22722, 22384, 21791, 217...
543836     [84987, 84991, 84988, 84992, 84378, 22965, 229...
543839                                [22207, 21731, 47590A]
543842                                               [22693]
543852     [22355, 21931, 22356, 85123A, 85175, 82484, 82...
543899     [C2

In [ ]:
transactions_va = va_data.groupby(['Country', 'month'])['StockCode'].apply(list)
min_sup = 0.45
frequentItemsets_va= list(apriori(transactions_va,min_support = min_sup))
for i in frequentItemsets_va:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'20719', '20726'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719', '20726'}), confidence=0.5, lift=1.0), OrderedStatistic(items_base=frozenset({'20719'}), items_add=frozenset({'20726'}), confidence=1.0, lift=2.0), OrderedStatistic(items_base=frozenset({'20726'}), items_add=frozenset({'20719'}), confidence=1.0, lift=2.0)])
yes
RelationRecord(items=frozenset({'20719', '20728'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'20719', '20728'}), confidence=0.5, lift=1.0), OrderedStatistic(items_base=frozenset({'20719'}), items_add=frozenset({'20728'}), confidence=1.0, lift=2.0), OrderedStatistic(items_base=frozenset({'20728'}), items_add=frozenset({'20719'}), confidence=1.0, lift=2.0)])
yes
RelationRecord(items=frozenset({'22383', '20719'}), support=0.5, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22383', '20719

##6.โดยจัดกลุ่มตามประเทศ+เดือน

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

In [ ]:
data_MC=data

In [ ]:
data_MC

In [ ]:
transactions_MC = data_MC.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [ ]:
transactions_MC

In [ ]:
!pip install apyori
from apyori import apriori

In [ ]:
apriori(transactions_MC,min_support = 0.53)


In [ ]:
list(apriori(transactions_MC,min_support = 0.53))


In [ ]:
frequentItemsets_MC = list(apriori(transactions_MC,min_support = 0.53))
frequentItemsets_MC

In [ ]:
min_sup=0.53
frequentItemsets_MC= list(apriori(transactions_MC,min_support =min_sup))
for i in frequentItemsets_MC:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
data_MC[data_MC['StockCode'] == '85123A'][['Month', 'Country','Description']]
